utils/context.go

```diff
@@ -0,0 +1,31 @@
+package utils
+
+import (
+	"context"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+const MetadataKey = "viam-metadata"
+
+func ContextWithMetadata(ctx context.Context) (context.Context, map[string][]string) {
+	md := make(map[string][]string)
+	ctx = context.WithValue(ctx, MetadataKey, md)
+	return ctx, md
+}
+
+func ContextWithMetadataUnaryClientInterceptor(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+	var header metadata.MD
+	opts = append(opts, grpc.Header(&header))
+	invoker(ctx, method, req, reply, cc, opts...)
+
+	md := ctx.Value(MetadataKey)
+	if mdMap, ok := md.(map[string][]string); ok {
+		for key, value := range header {
```

Comment:
[q] Could we consider just doing mdMap = header? Or are there other fields in mdMap we may want to not overwrite? Is this future proofing?

Reply:
If we set mdMap to header, we're just overwriting the value of mdMap in this specific function. The caller, which set mdMap on the context and has a reference to that map, will not see these changes. For example:

```go
package main

import "fmt"

func main() {
	myMap := make(map[string]string)
	fmt.Println("hello in main: ", myMap["hello"])
	addToMap(myMap)
	fmt.Println("hello in main: ", myMap["hello"])
}

func addToMap(someMap map[string]string) {
	someMap["hello"] = "world"
	anotherMap := map[string]string{
		"hello": "jeremy",
	}
	someMap = anotherMap
	fmt.Println("hello in addToMap: ", someMap["hello"])
}
```

output:
```
hello in main:  
hello in addToMap:  jeremy
hello in main:  world
```

Overwriting the metadata map on the context wouldnt work either because context.WithValue just wraps the context in a context.valueCtx, so we'd need a way to pass this modified context back to the caller, which isn't possible, and even if it were, modifying the context that the caller passed in is cleaner.